SOLR-15694 Concept of node roles and non-data nodes (#403)
apache/solr: solr/solr-ref-guide/src/node-roles.adoc

= Node Roles
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

> A node in Solr is usually capable of performing various types of operations, e.g. hosting replicas, performing indexing and querying on them, collection management tasks etc. However, if someone wants to setup a cluster where these functionalities are isolated to certain dedicated nodes, then this can be achieved leveraging the concept of node roles.

Reviewer: Also, can simplify this language.

```suggestion
A node in Solr is usually capable of performing various types of operations, e.g. hosting replicas, performing indexing and querying, collection management tasks, etc. To set up a cluster where these functions are isolated to certain dedicated nodes, we can use the concept of node roles.
```

Reviewer 2: Fixed, thanks.

== Definitions

=== Node role

> A role is a designation of a node that indicates that the node may perform a certain functionality that is governed by the role.

Reviewer: ```suggestion
A role is a designation for a node that indicates that the node may perform a certain function.
```

Reviewer 2: Fixed, thanks.

=== Mode
Every role has a list of modes under which a node can be. It can be simple (e.g. `["on", "off"]`) or more granular (e.g. `["allowed", "preferred", "disallowed"]`).

== Roles

In order to specify role(s) for a node, one needs to start a Solr node with the following parameter.

.Startup Parameter
[cols="1,2,1,1"] 
|===
|Parameter |Value |Required? | Default

|solr.node.roles
|Comma separated list of roles (in the format: `<role>:<mode>`) for this node.
Examples: `-Dsolr.node.roles=data:on,overseer:allowed` or `-Dsolr.node.roles=overseer:preferred`
|No

> |`data:on,overseer:allowed`

Reviewer: I don't understand what is wrong with this line?

// under the License.

A node in Solr is usually capable of performing various types of operations, e.g. hosting replicas, performing indexing and querying on them, collection management tasks etc. However, if someone wants to setup a cluster where these functionalities are isolated to certain dedicated nodes, then this can be achieved leveraging the concept of node roles.

== Definitions

=== Node role

A role is a designation of a node that indicates that the node may perform a certain functionality that is governed by the role.

=== Mode
Every role has a list of modes under which a node can be. It can be simple (e.g. `["on", "off"]`) or more granular (e.g. `["allowed", "preferred", "disallowed"]`).

== Roles

In order to specify role(s) for a node, one needs to start a Solr node with the following parameter.

.Startup Parameter
[cols="1,2,1,1"] 
|===
|Parameter |Value |Required? | Default

|solr.node.roles
|Comma separated list of roles (in the format: `<role>:<mode>`) for this node.
Examples: `-Dsolr.node.roles=data:on,overseer:allowed` or `-Dsolr.node.roles=overseer:preferred`
|No
|`data:on,overseer:allowed`
|===

[TIP]
====

> If a node has been started with no `solr.node.roles` parameter, it will be assumed to have the data role turned on and overseer role allowed on it. If you've never used roles before, you likely won't need to change anything in your startup parameters to accomodate the functionality associated with these roles.

Reviewer 3: typo: accommodate (two m)

Reviewer 2: Fixed, thanks.

====

> .Supported roles

Reviewer 3: Table should indicate default value for each role.

Reviewer 2: There is no concept of a default mode for a role.

[cols="1,1"] 
|===
|Role |Modes

|`data`
|on, off

|`overseer`
|allowed, preferred, disallowed
|===

=== `data` role
A node with this role (in mode "on") can host shards and replicas for collections.

=== `overseer` role
A node with this role can perform duties of an overseer node (unless mode is `disallowed`). When one or more nodes have the overseer role in `preferred` mode, the overseer leader will be elected from one of these nodes. In case no node is designated as a preferred overseer or no such node is live, the overseer leader will be elected from one of the nodes that have the overseer role in `allowed` mode. If all nodes that are designated with overseer role (allowed or preferred) are down, the cluster will be left without an overseer.

== Example usage

Sometimes, when the nodes in a cluster are under heavy querying or indexing load, the overseer leader node might be unable to perform collection management duties efficiently. It might be reasonable to have dedicated nodes to act as the overseer. Such an effect can be achieved as follows:

> * Most nodes (data nodes) in the cluster start with `-Dsolr.node.roles=data:on,overseer:allowed` (or with no parameter, since the default value for `solr.node.roles` is the same).

Reviewer 3: If you accept that the default should be per role (and I hope you do, it doesn't change cluster behavior with this PR but simplifies life for later), this sentence has to be rephrased (see comment line 41 of `NodeRoles`)

Reviewer 2: I don't think per role default is a sensible idea, I'd rather stick with per node default string. I had explained this in the SIP thread as follows:

```
   * User upgrades from Solr 9.0 to 9.1, where "ui" role has been introduced. Developers of "ui" role want it to be available for most users.
        - In Ilan's proposal, the developer chooses this in 9.1: ui: {modes: [on, off], default1: on, default2: on}. Now, user upgrading will see that UI is running on his two [dedicated] overseer nodes, and he's confused (because he explicitly specified what he wants)
        - In my proposal, the developer chooses ui: {modes: [on, off]}; default roles for those users who don't specify roles: "data:on, overseer:allowed, ui:on". Now, there are no surprises of implicit default. Users who don't use roles at all will get this functionality turned on, just as the developer wanted. Users who use roles will have to explicitly append "ui:on" to their roles string on their nodes during the upgrade (this tip will come from the upgrade notes).
```

In short, a user who specified "-Dsolr.node.roles=overseer:preferred" wants the node to only do overseer duties and nothing else, even if a Solr upgrade has a new role added that is supposed to be turned on by default on nodes where no role has been specified.

Reviewer 2: Here's the link to the quoted mail: https://lists.apache.org/thread/xn5z79hyb2sggwsqs41165krcj96hvdl

Reviewer 3: Having a per role default (`defaultIfAbsent`) as well as a default for all roles when no role is specified (default value for `solr.node.roles`) is obviously more complex than only per role default (understanding how a Solr node behaves requires looking at two places to understand default behavior).

One can assume that a Solr deployment where `solr.node.roles` is overridden on each node (an advanced or "expert" cluster config) would be careful on upgrades and will not fall into the "ui" issue you mention.

And in the general case, with roles independent of each other, picking a node to be (or not to be) overseer for example will have a side effect impact on all other roles (switching them from default value in `solr.node.roles` to per role default in `defaultIfAbsent`).
So a user not caring about roles in general but only wanting to exclude a given node from being Overseer would have to go look at the source code and do so on each code update and copy all defaults into its node's `solr.node.roles`, rather than just setting overseer to disallowed there.

Reviewer 4: Good points @murblanc. Especially I think this statement is true:

> One can assume that a Solr deployment where solr.node.roles is overridden on each node (an advanced or "expert" cluster config) would be careful on upgrades and will not fall into the "ui" issue you mention.

This seems to blow up in unnecessary configuration complexity, and optimizing for the needs of 5% of users, rather than 95%.

So why not start with an explicit WYSIWYG and KISS approach, and then in 9.x refine based on real-life upgrade needs? All these config facets and various defaults is a recipe for future bugs and confusion.

Reviewer 2: > Having a per role default (defaultIfAbsent) as well as a default for all roles when no role is specified (default value for solr.node.roles) is obviously more complex than only per role default (understanding how a Solr node behaves requires looking at two places to understand default behavior).

There is no per role default. What used to be called "defaultIfAbsent" is now renamed as "modeWhenRoleIsAbsent", and is used for internal purposes anyway.

> One can assume that a Solr deployment where solr.node.roles is overridden on each node (an advanced or "expert" cluster config) would be careful on upgrades and will not fall into the "ui" issue you mention.

I don't think we should make such an assumption. An upgrade should come with no surprises to the user, unless he/she explicitly wants to opt into a role introduced in the new release.
